src/main/java/com/smartcar/sdk/AuthClient.java

```diff
@@ -310,6 +312,101 @@ public AuthVehicleInfo build() {
     }
   }
 
+  /**
+   * A class that creates a custom AuthUrlConfig object, which is used
+   * when generating an authentication URL.
+   */
+  public static class AuthUrlConfig {
+    private boolean forcePrompt;
+    private String make;
+    private boolean singleSelect;
+    private String state;
+    
+    /** 
+     * A function that creates a map with query params and their values.
+     */
+    public HashMap getParams() {
+      HashMap<String, String> list = new HashMap<>();
+
+      list.put("single_select", this.singleSelect ? "true" : "false");
+      list.put("approval_prompt", this.forcePrompt ? "force" : "auto");
+
+      if(this.state != null) {
+          list.put("state", this.state);
+      }
+      if(this.make != null){
+        list.put("make", this.make);
+      }
+
+      return list;
+    }
+
+    public void setMake(String make) {
+      this.make = make;
+    }
+
+    public void setForcePrompt(boolean forcePrompt) {
+      this.forcePrompt = forcePrompt;
+    }
+
+    public void setSingleSelect(boolean singleSelect) {
+      this.singleSelect = singleSelect;
+    }
+
+    public void setState(String state) {
+      this.state = state;
+    }
+
+    /**
+    * Initializes a new AuthUrlconfig.
+    *
+    * @param forcePrompt whether to force the approval prompt to show every auth.
+    * @param make a vehicle make.
+    * @param singleSelect boolean which limits the number of vehicles a user can select to 1.
+    * @param state arbitrary string to be returned to the redirect URI.
+    */
+    public AuthUrlConfig(boolean forcePrompt, String make, boolean singleSelect, String state){
+      this.state = state;
+      this.forcePrompt = forcePrompt;
+      this.make = make;
+      this.singleSelect = singleSelect;
+    }
+
+    /**
+    * Initializes a new AuthUrlconfig.
+    */
+    public AuthUrlConfig() {
+        this(false, null, false, null);
+    }
+  }
+
+  /**
+   * Returns the assembled authentication URL.
+   *
+   * @param options an object containing query params to generate auth URL.
+   *
+   * @return the authentication URL
+   */
+  public String getAuthUrl(AuthUrlConfig options) {
+    HttpUrl.Builder urlBuilder = HttpUrl.parse(this.urlAuthorize).newBuilder()
+        .addQueryParameter("response_type", "code")
+        .addQueryParameter("client_id", this.clientId)
+        .addQueryParameter("redirect_uri", this.redirectUri);
+
+    if(this.scope != null) {
+      urlBuilder.addQueryParameter("scope", Utils.join(this.scope, " "));
```

Comment:
What is the behavior for an empty array of scopes?

Reply:
For empty arrays is going to add an empty string "" and add the query param as `scope=`. In api, even though we say it is required, if you don't pass anything, this will default to all the permissions.
I don't like this behavior but also do not sure if we should throw an error if its an empty scope.